src/ray/core_worker/transport/dependency_resolver.cc

```diff
@@ -105,52 +86,80 @@ void LocalDependencyResolver::ResolveDependencies(
   }
 
   // This is deleted when the last dependency fetch callback finishes.
-  std::shared_ptr<TaskState> state = std::make_shared<TaskState>(
-      task, std::move(local_dependencies), std::move(actor_dependences));
-  num_pending_ += 1;
+  const auto task_id = task.TaskId();
+  auto inserted = pending_tasks_.emplace(task_id,
+                                         new TaskState(task,
+                                                       std::move(local_dependencies),
+                                                       std::move(actor_dependences),
+                                                       on_complete));
+  RAY_CHECK(inserted.second);
+  auto &state = inserted.first->second;
 
   for (const auto &it : state->local_dependencies) {
     const ObjectID &obj_id = it.first;
     in_memory_store_.GetAsync(
-        obj_id, [this, state, obj_id, on_complete](std::shared_ptr<RayObject> obj) {
+        obj_id, [this, task_id, obj_id](std::shared_ptr<RayObject> obj) {
           RAY_CHECK(obj != nullptr);
-          bool complete = false;
+
+          std::unique_ptr<TaskState> resolved_task_state = nullptr;
           std::vector<ObjectID> inlined_dependency_ids;
           std::vector<ObjectID> contained_ids;
           {
             absl::MutexLock lock(&mu_);
+
+            auto it = pending_tasks_.find(task_id);
+            if (it == pending_tasks_.end()) {
+              return;
+            }
+            auto &state = it->second;
             state->local_dependencies[obj_id] = std::move(obj);
             if (--state->obj_dependencies_remaining == 0) {
               InlineDependencies(state->local_dependencies,
                                  state->task,
                                  &inlined_dependency_ids,
                                  &contained_ids);
               if (state->actor_dependencies_remaining == 0) {
-                complete = true;
-                num_pending_ -= 1;
+                resolved_task_state = std::move(state);
+                pending_tasks_.erase(it);
               }
             }
           }
+
           if (inlined_dependency_ids.size() > 0) {
             task_finisher_.OnTaskDependenciesInlined(inlined_dependency_ids,
                                                      contained_ids);
           }
-          if (complete) {
-            on_complete(state->status);
+          if (resolved_task_state) {
+            resolved_task_state->on_complete(resolved_task_state->status);
           }
         });
   }
 
   for (const auto &actor_id : state->actor_dependencies) {
     actor_creator_.AsyncWaitForActorRegisterFinish(
-        actor_id, [this, state, on_complete](const Status &status) {
-          if (!status.ok()) {
-            state->status = status;
+        actor_id, [this, task_id, on_complete](const Status &status) {
+          std::unique_ptr<TaskState> resolved_task_state = nullptr;
```

Comment:
there are some duplications for both `AsyncWaitForActorRegisterFinish` and `GetAsync`. Is is possible to unify them?

Reply:
Hmm I couldn't figure out a way to do this nicely.